moreorless/combined.py

```diff
@@ -0,0 +1,221 @@
+"""
+Computes the "combined diff format" used by git.
+
+This is one way of showing a 3+ way diff, such as when you have a tool that
+needs to show multiple diffs to a common ancestor (like input, expected
+output, and actual output); or a merge that combines multiple heads into one
+common descendant.  These are not intended to be applied.
+
+Instead of a single +/- character, there are n-1 of those characters at the
+start of the line, depending on which of the outputs (or inputs, in the case of
+a merge) that line is present in.
+
+If you're ever in a situation where you're analyzing a diff of diffs, this
+might be helpful in reducing confusion.
+
+Documented at https://git-scm.com/docs/diff-format#_combined_diff_format
+
+Example of two diffs vs a common ancestor:
+
+    $ python -m moreorless.combined a b c
+    --- a/file
+    +++ b/file
+    +++ b/file
+    @@@ -1,2 +1,2 +1,2 @@@
+      def main():
+    --    pass
+    +     print("b")
+     +    print("hi")
+"""
+
+from difflib import SequenceMatcher
+from typing import Dict, Iterable, List, Optional, Sequence, Union
+
+from .combined_types import Hunk, Line, Snip
+
+
+def combined_diff(
+    from_files: List[str],
+    to_files: List[str],
+    basename: str = "file",
+    from_filenames: Optional[List[str]] = None,
+    to_filenames: Optional[List[str]] = None,
+    context: int = 3,
+) -> str:
+    """
+    Returns a combined diff representing a multiway change.
+    """
+    if not from_filenames:
+        from_filenames = [f"a/{basename}"] * len(from_files)
+    if not to_filenames:
+        to_filenames = [f"b/{basename}"] * len(to_files)
+
+    assert len(from_files) == len(from_filenames)
+    assert len(to_files) == len(to_filenames)
+    assert len(from_files) == 1 or len(to_files) == 1
+    merge_mode = len(to_files) == 1
```

Comment:
There's ambiguity if you have len(from_files) == 1 and len(to_files) == 1, but I don't intend to call this unless one of them has more than one.  There's also probably errors if either of those lists is empty, but that too is a confusing usage.